### ISSUE TITLE
Issue with the function CountDecimalPlaces()

### ISSUE THREAD

**Issue by amrali-eg** [EarnForex/PositionSizeCalculator#22]
I suggest the following as improvement.

```mql5
#define PRINT( A) Print( #A + " = ", (A))

//+------------------------------------------------------------------+
//| Counts decimal places.                                           |
//+------------------------------------------------------------------+
//int CountDecimalPlaces(double number)
//{
//   // 100 as maximum length of number.
//   for (int i = 0; i < 100; i++)
//   {
//      if (MathAbs(MathRound(number) - number) / MathPow(10, i) <= FLT_EPSILON) return(i); // For some reason DBL_EPSILON doesn't work here.
//      number *= 10;
//   }
//   return(-1);
//}

int CountDecimalPlaces(double number)
{
   // 100 as maximum length of number.
   for (int i = 0; i < 100; i++)
   {
      double pwr = MathPow(10, i);
      if (MathRound(number * pwr) / pwr == number) return(i);
   }
   return(-1);
}
//+------------------------------------------------------------------+
//|                                                                  |
//+------------------------------------------------------------------+
void OnStart()
  {
   Print("======== Testing CountDecimalPlaces() ========");
   PRINT( CountDecimalPlaces(1) == 0 );
   PRINT( CountDecimalPlaces(0.1) == 1 );
   PRINT( CountDecimalPlaces(0.01) == 2 );
   PRINT( CountDecimalPlaces(0.001) == 3 );
   PRINT( CountDecimalPlaces(0.0001) == 4 );
   PRINT( CountDecimalPlaces(0.00001) == 5 );
   PRINT( CountDecimalPlaces(0.000001) == 6 );
   PRINT( CountDecimalPlaces(0.0000001) == 7 );
   PRINT( CountDecimalPlaces(0.00000001) == 8 );
   PRINT( CountDecimalPlaces(0.000000001) == 9 );
   PRINT( CountDecimalPlaces(0.0000000001) == 10 );
   PRINT( CountDecimalPlaces(0.00000000001) == 11 );
   PRINT( CountDecimalPlaces(0.000000000001) == 12 );
   PRINT( CountDecimalPlaces(0.0000000000001) == 13 );
   PRINT( CountDecimalPlaces(0.00000000000001) == 14 );
   PRINT( CountDecimalPlaces(0.000000000000001) == 15 );
   PRINT( CountDecimalPlaces(0.0000000000000001) == 16 );
   PRINT( CountDecimalPlaces(0.00000000000000001) == 17 );
   PRINT( CountDecimalPlaces(0.000000000000000001) == 18 );
   PRINT( CountDecimalPlaces(0.0000000000000000001) == 19 );

   PRINT( CountDecimalPlaces(1) == 0 );
   PRINT( CountDecimalPlaces(1.1) == 1 );
   PRINT( CountDecimalPlaces(1.12) == 2 );
   PRINT( CountDecimalPlaces(1.123) == 3 );
   PRINT( CountDecimalPlaces(1.1234) == 4 );
   PRINT( CountDecimalPlaces(1.12345) == 5 );
   PRINT( CountDecimalPlaces(1.123456) == 6 );
   PRINT( CountDecimalPlaces(1.1234567) == 7 );
   PRINT( CountDecimalPlaces(1.12345678) == 8 );
   PRINT( CountDecimalPlaces(1.123456789) == 9 );
   PRINT( CountDecimalPlaces(1.1234567891) == 10 );
   PRINT( CountDecimalPlaces(1.12345678912) == 11 );
   PRINT( CountDecimalPlaces(1.123456789123) == 12 );
   PRINT( CountDecimalPlaces(1.1234567891234) == 13 );
   PRINT( CountDecimalPlaces(1.12345678912345) == 14 );
   PRINT( CountDecimalPlaces(0.012345678912345) == 15 );
   PRINT( CountDecimalPlaces(0.0123456789123456) == 16 );
   PRINT( CountDecimalPlaces(0.00123456789123456) == 17 );
   PRINT( CountDecimalPlaces(0.000123456789123456) == 18 );
   PRINT( CountDecimalPlaces(0.0000123456789123456) == 19 );

   PRINT( CountDecimalPlaces(1.12) == 2 );
   PRINT( CountDecimalPlaces(1.14) == 2 );
   PRINT( CountDecimalPlaces(1.36) == 2 );
   PRINT( CountDecimalPlaces(1.57) == 2 );
   PRINT( CountDecimalPlaces(1.93) == 2 );
   PRINT( CountDecimalPlaces(1.118) == 3 );
   PRINT( CountDecimalPlaces(1.378) == 3 );
   PRINT( CountDecimalPlaces(1.757) == 3 );
   PRINT( CountDecimalPlaces(1.816) == 3 );
   PRINT( CountDecimalPlaces(90.268) == 3 );
   PRINT( CountDecimalPlaces(1.0778) == 4 );
   PRINT( CountDecimalPlaces(1.0954) == 4 );
   PRINT( CountDecimalPlaces(1.2749) == 4 );
   PRINT( CountDecimalPlaces(0.85919) == 5 );
   PRINT( CountDecimalPlaces(1.09545) == 5 );
   PRINT( CountDecimalPlaces(1.26008) == 5 );
   PRINT( CountDecimalPlaces(1.27481) == 5 );
   PRINT( CountDecimalPlaces(1.27483) == 5 );
   PRINT( CountDecimalPlaces(1.27571) == 5 );
   PRINT( CountDecimalPlaces(1.30418) == 5 );
   PRINT( CountDecimalPlaces(1.42147) == 5 );
   PRINT( CountDecimalPlaces(1.54289) == 5 );
   PRINT( CountDecimalPlaces(1.56849) == 5 );
   PRINT( CountDecimalPlaces(1.61465) == 5 );
   PRINT( CountDecimalPlaces(1.69292) == 5 );
  }
//+------------------------------------------------------------------+
```

**Comment by EarnForex** [comment]
Could you please describe your suggestion in more detail?

**Comment by amrali-eg** [comment]
The function is able to count decimal places more accurately. I provided some test cases.
Best regards,

**Comment by EarnForex** [comment]
OK, thank you! I will apply it in the next version.

**Comment by EarnForex** [comment]
Fixed in 2.40.
